src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -409,5 +559,118 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::KDF(const uint8_t * ikm, const size_t
+    integer_length = EmitDerIntegerFromRaw(&raw_sig[fe_length_bytes], fe_length_bytes, cursor, remaining);
+    VerifyOrReturnError(integer_length != 0, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(integer_length <= remaining, CHIP_ERROR_INTERNAL);
+    integers_length += integer_length;
+
+    // We only support outputs that would use 1 or 2 bytes of DER length after the SEQUENCE tag
+    VerifyOrReturnError(integers_length <= UINT8_MAX, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // We now know the length of both variable sized integers in the sequence, so we
+    // can write the tag and length.
+    BufferWriter writer(out_asn1_sig, static_cast<uint16_t>(out_asn1_sig_length));
+
+    // Put SEQUENCE tag
+    writer.Put(kSeqTag);
+
+    // Put the length over 1 or two bytes depending on case
+    if (integers_length > 127u)
+    {
+        writer.Put(static_cast<uint8_t>(0x80u | 1u)); // Length is extended length, over 1 subsequent byte
+        writer.Put(static_cast<uint8_t>(integers_length));
+    }
+    else
+    {
+        // Length is directly in the first byte with MSB clear if <= 127.
+        writer.Put(static_cast<uint8_t>(integers_length));
+    }
+
+    // Put the contents of the integers previously serialized in the buffer.
+    // The writer.Put is memmove-safe, so the shifting will happen from the read
+    // of the same buffer where the write is taking place.
+    writer.Put(&out_asn1_sig[kMinOverhead], integers_length);
+
+    size_t actually_written = 0;
+    if (writer.Fit(actually_written))
+    {
+        out_asn1_sig_actual_length = actually_written;
+        return CHIP_NO_ERROR;
+    }
+
+    return CHIP_ERROR_BUFFER_TOO_SMALL;
+}
+
+CHIP_ERROR EcdsaAsn1SignatureToRaw(size_t fe_length_bytes, const uint8_t * asn1_sig, size_t asn1_sig_length, uint8_t * out_raw_sig,
+                                   size_t out_raw_sig_length)
+{
+    VerifyOrReturnError(fe_length_bytes > 0, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(asn1_sig != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    constexpr size_t kMinOverhead = 1u + 2u; // 1 byte tag, (1 byte length, 1 byte data) or 3 bytes length
+    VerifyOrReturnError(asn1_sig_length > kMinOverhead, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Output raw signature is <r,s> both of which are of fe_length_bytes (see SEC1).
+    VerifyOrReturnError(out_raw_sig_length >= (2u * fe_length_bytes), CHIP_ERROR_BUFFER_TOO_SMALL);
+    VerifyOrReturnError(out_raw_sig != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    chip::Encoding::LittleEndian::Reader reader(asn1_sig, static_cast<uint16_t>(asn1_sig_length));
+
+    // Make sure we have a starting Sequence
+    uint8_t tag = 0;
+    VerifyOrReturnError(reader.Read8(&tag).StatusCode() == CHIP_NO_ERROR, reader.StatusCode());
+    VerifyOrReturnError(tag == kSeqTag, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Read length of sequence
+    uint8_t tag_len   = 0;
+    CHIP_ERROR status = ReadDerLength(reader, tag_len);
```

Comment:
usage of status feels odd. I believe we could avoid explicitly declaring it when using return macros if we never do anything with it except returning:

```cpp
ReturnErrorOnFailure(ReadDerLength(reader, tag_len));
VerifyOrReturnError(tag_len == reader.Remaining(), CHIP_ERROR_INVALID_ARGUMENT);
...
ReturnErrorOnFailure(ReadDerUnsignedIntegerIntoRaw(reader, raw_cursor, fe_length_bytes));
...
```

Reply:
Done